From 197c19551c277fc59fc3e3927fadcf73c9ddfe0f Mon Sep 17 00:00:00 2001
From: Cem Tenruh <c.tenruh@phytec.de>
Date: Tue, 15 Nov 2022 08:04:22 +0100
Subject: [PATCH] BACKPORT: opencv: Allow building against 4.6.x

duplicated this patch from gstreamer upstream repository:
fae27c87de36c548d3ae1aad290806f3424c5c54
could not cherry-pick because of path inconsistencies.
Replace the broken version checks with one modeled after
`GLIB_CHECK_VERSION`.

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/2558>

Signed-off-by: Cem Tenruh <c.tenruh@phytec.de>
---
 ext/opencv/gstcvtracker.cpp | 14 +++++++-------
 ext/opencv/gstcvtracker.h   | 10 ++++++++--
 ext/opencv/meson.build      |  2 +-
 3 files changed, 16 insertions(+), 10 deletions(-)

diff --git a/ext/opencv/gstcvtracker.cpp b/ext/opencv/gstcvtracker.cpp
index 492c9cf9b8..0c5dfda35f 100644
--- a/ext/opencv/gstcvtracker.cpp
+++ b/ext/opencv/gstcvtracker.cpp
@@ -236,7 +236,7 @@ gst_cvtracker_init (GstCVTracker * filter)
   filter->y = DEFAULT_PROP_INITIAL_Y;
   filter->width = DEFAULT_PROP_INITIAL_WIDTH;
   filter->height = DEFAULT_PROP_INITIAL_HEIGHT;
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
   filter->tracker = cv::legacy::upgradeTrackingAPI(
       cv::legacy::TrackerMedianFlow::create());
 #else
@@ -286,7 +286,7 @@ create_cvtracker (GstCVTracker * filter)
 {
   switch (filter->algorithm) {
     case GST_OPENCV_TRACKER_ALGORITHM_BOOSTING:
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
       filter->tracker = cv::legacy::upgradeTrackingAPI(
           cv::legacy::TrackerBoosting::create());
 #else
@@ -300,7 +300,7 @@ create_cvtracker (GstCVTracker * filter)
       filter->tracker = cv::TrackerKCF::create ();
       break;
     case GST_OPENCV_TRACKER_ALGORITHM_MEDIANFLOW:
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
       filter->tracker = cv::legacy::upgradeTrackingAPI(
           cv::legacy::TrackerMedianFlow::create());
 #else
@@ -311,7 +311,7 @@ create_cvtracker (GstCVTracker * filter)
       filter->tracker = cv::TrackerMIL::create ();
       break;
     case GST_OPENCV_TRACKER_ALGORITHM_MOSSE:
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
       filter->tracker = cv::legacy::upgradeTrackingAPI(
           cv::legacy::TrackerMOSSE::create());
 #else
@@ -319,7 +319,7 @@ create_cvtracker (GstCVTracker * filter)
 #endif
       break;
     case GST_OPENCV_TRACKER_ALGORITHM_TLD:
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
       filter->tracker = cv::legacy::upgradeTrackingAPI(
           cv::legacy::TrackerTLD::create());
 #else
@@ -369,7 +369,7 @@ gst_cvtracker_transform_ip (GstOpencvVideoFilter * base,
   GstMessage *msg;
 
   if (filter->roi.empty ()) {
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
     filter->roi = new (cv::Rect);
 #else
     filter->roi = new (cv::Rect2d);
@@ -381,7 +381,7 @@ gst_cvtracker_transform_ip (GstOpencvVideoFilter * base,
     create_cvtracker (filter);
     filter->tracker->init (img, *filter->roi);
   } else if (filter->tracker->update (img, *filter->roi)) {
-#if (!(CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1))
+#if !GST_OPENCV_CHECK_VERSION(4, 5, 1)
     /* Round values to avoid inconsistencies depending on the OpenCV version. */
     filter->roi->x = cvRound (filter->roi->x);
     filter->roi->y = cvRound (filter->roi->y);
diff --git a/ext/opencv/gstcvtracker.h b/ext/opencv/gstcvtracker.h
index d3531efbc5..7403677d59 100644
--- a/ext/opencv/gstcvtracker.h
+++ b/ext/opencv/gstcvtracker.h
@@ -53,7 +53,13 @@
 #include <opencv2/core.hpp>
 #include <opencv2/imgproc.hpp>
 #include <opencv2/tracking.hpp>
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#define GST_OPENCV_CHECK_VERSION(major,minor,revision) \
+  (CV_VERSION_MAJOR > (major) || \
+   (CV_VERSION_MAJOR == (major) && CV_VERSION_MINOR > (minor)) || \
+   (CV_VERSION_MAJOR == (major) && CV_VERSION_MINOR == (minor) && \
+    CV_VERSION_REVISION >= (revision)))
+
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
 #include <opencv2/tracking/tracking_legacy.hpp>
 #endif
 
@@ -87,7 +93,7 @@ struct _GstCVTracker
   gboolean post_debug_info;
 
   cv::Ptr<cv::Tracker> tracker;
-#if CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 5 && CV_VERSION_REVISION >= 1
+#if GST_OPENCV_CHECK_VERSION(4, 5, 1)
   cv::Ptr<cv::Rect> roi;
 #else
   cv::Ptr<cv::Rect2d> roi;
diff --git a/ext/opencv/meson.build b/ext/opencv/meson.build
index b0bd54e803..4be2262c59 100644
--- a/ext/opencv/meson.build
+++ b/ext/opencv/meson.build
@@ -74,7 +74,7 @@ if opencv_found
 endif
 
 if not opencv_found
-  opencv_dep = dependency('opencv4', version : ['>= 4.0.0', '< 4.6.0'], required : false)
+  opencv_dep = dependency('opencv4', version : ['>= 4.0.0', '< 4.7.0'], required : false)
   opencv_found = opencv_dep.found()
   if opencv_found
     foreach h : libopencv4_headers
-- 
2.25.1

